give hints when an assertion value is None instead of a boolean

sometimes beginners invoke code like `assert something_that_asserts_and_returns_none` and end up puzzled,

for reference see https://stackoverflow.com/questions/48152560/pytest-assert-has-calls-usage/48206730?noredirect=1#comment83395353_48206730

i believe it would be a nice and simple addition to give a hint in the case that a assertion value was None and this might be a misstake,

for clarity `None` should be tested with `x is None` or `x is not None`

Hi @jrbenny35, thanks for reaching out!

I don't know the intricacies of the assertion rewriting myself, but you can start digging around the `_pytest/assertion` directory. 

For example currently a warning is issued when an `assert` is used with a `tuple` [here](https://github.com/pytest-dev/pytest/blob/71527072801b6d83db168996529be1e76ce96bc4/_pytest/assertion/rewrite.py#L733-L736). 

While it is possible to know that an `assert` statement is receiving a tuple at *parsing* time, to realize if an arbitrary expression is returning `None` we need to check this at runtime.

I would start by adding a breakpoint in `AssertionRewriter.visit_Assert` while debugging a test like this:

```python
def x(): return None

def test():
    assert x()
```

To try to figure out how pytest generates this message:

```
test_asserts.py:5 (test)
def test():
>       assert x()
E       assert None
E        +  where None = x()
```

Hope this helps! 👍

--- a/changelog/3191.feature.rst
+++ b/changelog/3191.feature.rst
@@ -1,4 +1,4 @@
-A warning is now issued when assertions are made directly against ``None``.
+A warning is now issued when assertions are made for ``None``.
 
 This is a common source of confusion among new users, which write::
 
@@ -14,4 +14,4 @@ This warning will not be issued when ``None`` is explicitly checked. An assertio
 
     assert variable is None
 
-will not issue the warning
+will not issue the warning.

--- a/changelog/3191.feature.rst
+++ b/changelog/3191.feature.rst
@@ -1,1 +1,11 @@
-Add warning when asserting ``None`` directly
+A warning is now issued when assertions are made directly against ``None``.
+
+This is a common source of confusion among new users, which write::
+
+    assert mocked_object.assert_called_with(3, 4, 5, key='value')
+    
+When they should write::     
+
+    mocked_object.assert_called_with(3, 4, 5, key='value')
+
+Because the ``assert_called_with`` method of mock objects already executes an assertion.

--- a/changelog/3191.feature.rst
+++ b/changelog/3191.feature.rst
@@ -10,7 +10,8 @@ When they should write::
 
 Because the ``assert_called_with`` method of mock objects already executes an assertion.
 
-This warning will not be issued when ``None`` is explicitly checked
-    assert none_returning_fun() is None
+This warning will not be issued when ``None`` is explicitly checked. An assertion like::
+
+    assert variable is None
 
 will not issue the warning

--- a/changelog/3191.feature.rst
+++ b/changelog/3191.feature.rst
@@ -1,0 +1,1 @@
+Add warning when asserting ``None`` directly

--- a/changelog/3191.feature.rst
+++ b/changelog/3191.feature.rst
@@ -3,9 +3,14 @@ A warning is now issued when assertions are made directly against ``None``.
 This is a common source of confusion among new users, which write::
 
     assert mocked_object.assert_called_with(3, 4, 5, key='value')
-    
-When they should write::     
+
+When they should write::
 
     mocked_object.assert_called_with(3, 4, 5, key='value')
 
 Because the ``assert_called_with`` method of mock objects already executes an assertion.
+
+This warning will not be issued when ``None`` is explicitly checked
+    assert none_returning_fun() is None
+
+will not issue the warning

--- a/src/_pytest/assertion/rewrite.py
+++ b/src/_pytest/assertion/rewrite.py
@@ -841,12 +841,13 @@ class AssertionRewriter(ast.NodeVisitor):
         self.push_format_context()
         # Rewrite assert into a bunch of statements.
         top_condition, explanation = self.visit(assert_.test)
-        # Check if directly asserting None, in order to warn [Issue #3191]
-        self.statements.append(
-            self.warn_about_none_ast(
-                top_condition, module_path=self.module_path, lineno=assert_.lineno
+        # If in a test module, check if directly asserting None, in order to warn [Issue #3191]
+        if self.module_path is not None:
+            self.statements.append(
+                self.warn_about_none_ast(
+                    top_condition, module_path=self.module_path, lineno=assert_.lineno
+                )
             )
-        )
         # Create failure message.
         body = self.on_failure
         negation = ast.UnaryOp(ast.Not(), top_condition)

--- a/src/_pytest/assertion/rewrite.py
+++ b/src/_pytest/assertion/rewrite.py
@@ -878,16 +878,10 @@ class AssertionRewriter(ast.NodeVisitor):
         return self.statements
 
     def warn_about_none_ast(self, node, module_path, lineno):
-        """Returns an ast warning if node is None with the following statement:
-        if node is None:
-            from _pytest.warning_types import PytestWarning
-            import warnings
-            warnings.warn_explicit(
-                PytestWarning('assertion the value None, Please use "assert is None"'),
-                category=None,
-                filename=str,
-                lineno=node.lineno,
-            )
+        """
+        Returns an ast issuing a warning if the value of node is `None`
+        This is used to warn the user when asserting a function that asserts internally.
+        See issue #3191 for more details
         """
 
         # using parse because it's different between py2 py3

--- a/src/_pytest/assertion/rewrite.py
+++ b/src/_pytest/assertion/rewrite.py
@@ -885,8 +885,7 @@ class AssertionRewriter(ast.NodeVisitor):
             warnings.warn_explicit(
                 PytestWarning('assertion the value None, Please use "assert is None"'),
                 category=None,
-                # filename=str(self.module_path),
-                filename=__file__
+                filename=str,
                 lineno=node.lineno,
             )
         """
@@ -905,7 +904,7 @@ warn_explicit(
     lineno={lineno},
 )
             """.format(
-                filename=str(module_path), lineno=lineno
+                filename=module_path.strpath, lineno=lineno
             )
         ).body
         return ast.If(val_is_none, send_warning, [])

--- a/src/_pytest/assertion/rewrite.py
+++ b/src/_pytest/assertion/rewrite.py
@@ -891,30 +891,24 @@ class AssertionRewriter(ast.NodeVisitor):
             )
         """
 
-        warning_msg = ast.Str(
-            'Asserting the value None directly, Please use "assert is None" to eliminate ambiguity'
-        )
-        AST_NONE = ast.NameConstant(None)
+        # using parse because it's different between py2 py3
+        AST_NONE = ast.parse("None").body[0].value
         val_is_none = ast.Compare(node, [ast.Is()], [AST_NONE])
-        import_warnings = ast.ImportFrom(
-            module="warnings", names=[ast.alias("warn_explicit", None)], level=0
-        )
-        import_pytest_warning = ast.ImportFrom(
-            module="pytest", names=[ast.alias("PytestWarning", None)], level=0
-        )
-        pytest_warning = ast_Call_helper("PytestWarning", warning_msg)
-        # This won't work because this isn't the same "self" as an AssertionRewriter!
-        # ast_filename = improved_ast_Call('str',ast.Attribute('self','module_path',ast.Load).module_path)
-        warn = ast_Call_helper(
-            "warn_explicit",
-            pytest_warning,
-            category=AST_NONE,
-            filename=ast.Str(str(module_path)),
-            lineno=ast.Num(lineno),
-        )
-        return ast.If(
-            val_is_none, [import_warnings, import_pytest_warning, ast.Expr(warn)], []
-        )
+        send_warning = ast.parse(
+            """
+from _pytest.warning_types import PytestWarning
+from warnings import warn_explicit
+warn_explicit(
+    PytestWarning('assertion the value None, Please use "assert is None"'),
+    category=None,
+    filename='{filename}',
+    lineno={lineno},
+)
+            """.format(
+                filename=str(module_path), lineno=lineno
+            )
+        ).body
+        return ast.If(val_is_none, send_warning, [])
 
     def visit_Name(self, name):
         # Display the repr of the name if it's a local variable or

--- a/src/_pytest/assertion/rewrite.py
+++ b/src/_pytest/assertion/rewrite.py
@@ -880,12 +880,13 @@ class AssertionRewriter(ast.NodeVisitor):
 
     def warn_about_none_ast(self, node, module_path, lineno):
         """
-        Returns an ast issuing a warning if the value of node is `None`
-        This is used to warn the user when asserting a function that asserts internally.
-        See issue #3191 for more details
+        Returns an AST issuing a warning if the value of node is `None`.
+        This is used to warn the user when asserting a function that asserts
+        internally already.
+        See issue #3191 for more details.
         """
 
-        # using parse because it's different between py2 py3
+        # Using parse because it is different between py2 and py3.
         AST_NONE = ast.parse("None").body[0].value
         val_is_none = ast.Compare(node, [ast.Is()], [AST_NONE])
         send_warning = ast.parse(
@@ -893,7 +894,7 @@ class AssertionRewriter(ast.NodeVisitor):
 from _pytest.warning_types import PytestWarning
 from warnings import warn_explicit
 warn_explicit(
-    PytestWarning('assertion the value None, Please use "assert is None"'),
+    PytestWarning('asserting the value None, please use "assert is None"'),
     category=None,
     filename={filename!r},
     lineno={lineno},

--- a/src/_pytest/assertion/rewrite.py
+++ b/src/_pytest/assertion/rewrite.py
@@ -51,6 +51,19 @@ else:
         return ast.Call(a, b, c, None, None)
 
 
+def ast_Call_helper(func_name, *args, **kwargs):
+    """
+    func_name: str
+    args: Iterable[ast.expr]
+    kwargs: Dict[str,ast.expr]
+    """
+    return ast.Call(
+        ast.Name(func_name, ast.Load()),
+        list(args),
+        [ast.keyword(key, val) for key, val in kwargs.items()],
+    )
+
+
 class AssertionRewritingHook(object):
     """PEP302 Import hook which rewrites asserts."""
 
@@ -828,6 +841,12 @@ class AssertionRewriter(ast.NodeVisitor):
         self.push_format_context()
         # Rewrite assert into a bunch of statements.
         top_condition, explanation = self.visit(assert_.test)
+        # Check if directly asserting None, in order to warn [Issue #3191]
+        self.statements.append(
+            self.warn_about_none_ast(
+                top_condition, module_path=self.module_path, lineno=assert_.lineno
+            )
+        )
         # Create failure message.
         body = self.on_failure
         negation = ast.UnaryOp(ast.Not(), top_condition)
@@ -857,6 +876,45 @@ class AssertionRewriter(ast.NodeVisitor):
         for stmt in self.statements:
             set_location(stmt, assert_.lineno, assert_.col_offset)
         return self.statements
+
+    def warn_about_none_ast(self, node, module_path, lineno):
+        """Returns an ast warning if node is None with the following statement:
+        if node is None:
+            from _pytest.warning_types import PytestWarning
+            import warnings
+            warnings.warn_explicit(
+                PytestWarning('assertion the value None, Please use "assert is None"'),
+                category=None,
+                # filename=str(self.module_path),
+                filename=__file__
+                lineno=node.lineno,
+            )
+        """
+
+        warning_msg = ast.Str(
+            'Asserting the value None directly, Please use "assert is None" to eliminate ambiguity'
+        )
+        AST_NONE = ast.NameConstant(None)
+        val_is_none = ast.Compare(node, [ast.Is()], [AST_NONE])
+        import_warnings = ast.ImportFrom(
+            module="warnings", names=[ast.alias("warn_explicit", None)], level=0
+        )
+        import_pytest_warning = ast.ImportFrom(
+            module="pytest", names=[ast.alias("PytestWarning", None)], level=0
+        )
+        pytest_warning = ast_Call_helper("PytestWarning", warning_msg)
+        # This won't work because this isn't the same "self" as an AssertionRewriter!
+        # ast_filename = improved_ast_Call('str',ast.Attribute('self','module_path',ast.Load).module_path)
+        warn = ast_Call_helper(
+            "warn_explicit",
+            pytest_warning,
+            category=AST_NONE,
+            filename=ast.Str(str(module_path)),
+            lineno=ast.Num(lineno),
+        )
+        return ast.If(
+            val_is_none, [import_warnings, import_pytest_warning, ast.Expr(warn)], []
+        )
 
     def visit_Name(self, name):
         # Display the repr of the name if it's a local variable or

--- a/src/_pytest/assertion/rewrite.py
+++ b/src/_pytest/assertion/rewrite.py
@@ -901,7 +901,7 @@ from warnings import warn_explicit
 warn_explicit(
     PytestWarning('assertion the value None, Please use "assert is None"'),
     category=None,
-    filename='{filename}',
+    filename={filename!r},
     lineno={lineno},
 )
             """.format(

--- a/testing/test_warnings.py
+++ b/testing/test_warnings.py
@@ -623,6 +623,8 @@ def test_removed_in_pytest4_warning_as_error(testdir, change_default):
     else:
         assert change_default in ("ini", "cmdline")
         result.stdout.fnmatch_lines(["* 1 passed in *"])
+
+
 class TestAssertionWarnings:
     @staticmethod
     def assert_result_warns(result):
@@ -660,7 +662,6 @@ class TestAssertionWarnings:
         result = testdir.runpytest()
         self.assert_result_warns(result)
 
-    @pytest.mark.xfail(strict=True)
     def test_assert_is_none_no_warn(self, testdir):
         """Tests a more simple case of `test_none_function_warns` where `assert None` is explicitly called"""
         testdir.makepyfile(
@@ -673,10 +674,9 @@ class TestAssertionWarnings:
             """
         )
         result = testdir.runpytest()
-        self.assert_result_warns(result)
+        result.stdout.fnmatch_lines(["*1 passed in*"])
 
-    @pytest.mark.xfail(strict=True)
     def test_false_function_no_warn(self, testdir):
         self.create_file(testdir, False)
         result = testdir.runpytest()
-        self.assert_result_warns(result)
+        result.stdout.fnmatch_lines(["*1 failed in*"])

--- a/testing/test_warnings.py
+++ b/testing/test_warnings.py
@@ -624,6 +624,10 @@ def test_removed_in_pytest4_warning_as_error(testdir, change_default):
         assert change_default in ("ini", "cmdline")
         result.stdout.fnmatch_lines(["* 1 passed in *"])
 class TestAssertionWarnings:
+    @staticmethod
+    def result_warns(result):
+        return result.stdout.fnmatch_lines(["*PytestWarning*"])
+
     def test_tuple_warning(self, testdir):
         testdir.makepyfile(
             """
@@ -631,10 +635,11 @@ class TestAssertionWarnings:
                 assert (1,2)
             """
         )
-        with pytest.warns(pytest.PytestWarning):
-            testdir.runpytest_subprocess()
+        result = testdir.runpytest()
+        assert self.result_warns(result)
 
-    def create_file(self, testdir, return_none):
+    @staticmethod
+    def create_file(testdir, return_none):
         testdir.makepyfile(
             """
             def foo(return_none):
@@ -652,9 +657,27 @@ class TestAssertionWarnings:
 
     def test_none_function_warns(self, testdir):
         self.create_file(testdir, True)
-        with pytest.warns(pytest.PytestWarning):
-            testdir.runpytest_subprocess()
+        result = testdir.runpytest()
+        assert self.result_warns(result)
+
+    def test_assert_is_none_no_warn(self, testdir):
+        """Tests a more simple case of `test_none_function_warns` where `assert None` is explicitly called"""
+        testdir.makepyfile(
+            """
+            def foo(return_none):
+                if return_none:
+                    return None
+                else:
+                    return False
+
+            def test_foo():
+                assert foo(True) is None
+            """
+        )
+        result = testdir.runpytest()
+        assert not self.result_warns(result)
 
     def test_false_function_no_warn(self, testdir):
         self.create_file(testdir, False)
-        testdir.runpytest_subprocess("-W error:PytestWarning")
+        result = testdir.runpytest()
+        assert not self.result_warns(result)

--- a/testing/test_warnings.py
+++ b/testing/test_warnings.py
@@ -623,3 +623,38 @@ def test_removed_in_pytest4_warning_as_error(testdir, change_default):
     else:
         assert change_default in ("ini", "cmdline")
         result.stdout.fnmatch_lines(["* 1 passed in *"])
+class TestAssertionWarnings:
+    def test_tuple_warning(self, testdir):
+        testdir.makepyfile(
+            """
+            def test_foo():
+                assert (1,2)
+            """
+        )
+        with pytest.warns(pytest.PytestWarning):
+            testdir.runpytest_subprocess()
+
+    def create_file(self, testdir, return_none):
+        testdir.makepyfile(
+            """
+            def foo(return_none):
+                if return_none:
+                    return None
+                else:
+                    return False
+
+            def test_foo():
+                assert foo({return_none})
+            """.format(
+                return_none=return_none
+            )
+        )
+
+    def test_none_function_warns(self, testdir):
+        self.create_file(testdir, True)
+        with pytest.warns(pytest.PytestWarning):
+            testdir.runpytest_subprocess()
+
+    def test_false_function_no_warn(self, testdir):
+        self.create_file(testdir, False)
+        testdir.runpytest_subprocess("-W error:PytestWarning")

--- a/testing/test_warnings.py
+++ b/testing/test_warnings.py
@@ -625,8 +625,8 @@ def test_removed_in_pytest4_warning_as_error(testdir, change_default):
         result.stdout.fnmatch_lines(["* 1 passed in *"])
 class TestAssertionWarnings:
     @staticmethod
-    def result_warns(result):
-        return result.stdout.fnmatch_lines(["*PytestWarning*"])
+    def assert_result_warns(result):
+        result.stdout.fnmatch_lines(["*PytestWarning*"])
 
     def test_tuple_warning(self, testdir):
         testdir.makepyfile(
@@ -636,7 +636,7 @@ class TestAssertionWarnings:
             """
         )
         result = testdir.runpytest()
-        assert self.result_warns(result)
+        self.assert_result_warns(result)
 
     @staticmethod
     def create_file(testdir, return_none):
@@ -658,26 +658,25 @@ class TestAssertionWarnings:
     def test_none_function_warns(self, testdir):
         self.create_file(testdir, True)
         result = testdir.runpytest()
-        assert self.result_warns(result)
+        self.assert_result_warns(result)
 
+    @pytest.mark.xfail(strict=True)
     def test_assert_is_none_no_warn(self, testdir):
         """Tests a more simple case of `test_none_function_warns` where `assert None` is explicitly called"""
         testdir.makepyfile(
             """
-            def foo(return_none):
-                if return_none:
-                    return None
-                else:
-                    return False
+            def foo():
+                return None
 
             def test_foo():
-                assert foo(True) is None
+                assert foo() is None
             """
         )
         result = testdir.runpytest()
-        assert not self.result_warns(result)
+        self.assert_result_warns(result)
 
+    @pytest.mark.xfail(strict=True)
     def test_false_function_no_warn(self, testdir):
         self.create_file(testdir, False)
         result = testdir.runpytest()
-        assert not self.result_warns(result)
+        self.assert_result_warns(result)

--- a/testing/test_warnings.py
+++ b/testing/test_warnings.py
@@ -627,8 +627,8 @@ def test_removed_in_pytest4_warning_as_error(testdir, change_default):
 
 class TestAssertionWarnings:
     @staticmethod
-    def assert_result_warns(result):
-        result.stdout.fnmatch_lines(["*PytestWarning*"])
+    def assert_result_warns(result, msg):
+        result.stdout.fnmatch_lines(["*PytestWarning: %s*" % msg])
 
     def test_tuple_warning(self, testdir):
         testdir.makepyfile(
@@ -638,7 +638,9 @@ class TestAssertionWarnings:
             """
         )
         result = testdir.runpytest()
-        self.assert_result_warns(result)
+        self.assert_result_warns(
+            result, "assertion is always true, perhaps remove parentheses?"
+        )
 
     @staticmethod
     def create_file(testdir, return_none):
@@ -660,10 +662,11 @@ class TestAssertionWarnings:
     def test_none_function_warns(self, testdir):
         self.create_file(testdir, True)
         result = testdir.runpytest()
-        self.assert_result_warns(result)
+        self.assert_result_warns(
+            result, 'asserting the value None, please use "assert is None"'
+        )
 
     def test_assert_is_none_no_warn(self, testdir):
-        """Tests a more simple case of `test_none_function_warns` where `assert None` is explicitly called"""
         testdir.makepyfile(
             """
             def foo():

